wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt

```diff
@@ -0,0 +1,217 @@
+import org.jetbrains.kotlin.gradle.plugin.KotlinBasePluginWrapper
+import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
+import java.io.File
+import java.net.URI
+import javax.inject.Inject
+
+class WirePlugin @Inject constructor(
+  private val sourceDirectorySetFactory: SourceDirectorySetFactory
+) : Plugin<Project> {
+  private var kotlin = false
+  private var java = false
+
+  override fun apply(project: Project) {
+    val logger = project.logger
+
+    val extension = project.extensions.create(
+        "wire", WireExtension::class.java, project, sourceDirectorySetFactory
+    )
+
+    project.plugins.all {
+      logger.debug("plugin: $it")
+      when (it) {
+        is JavaBasePlugin -> {
+          java = true
+        }
+        is KotlinBasePluginWrapper -> {
+          kotlin = true
+        }
+      }
+    }
+
+    if (!kotlin && !java) {
+      throw IllegalArgumentException(
+          "The Wire Gradle plugin requires either the Java or Kotlin plugin to be applied prior to its being applied."
+      )
+    }
+
+    project.afterEvaluate {
+      val ssc = it.property("sourceSets") as SourceSetContainer
+      if (logger.isDebugEnabled) {
+        ssc.forEach {
+          logger.debug("source set: ${it.name}")
+        }
+      }
+
+      when {
+        kotlin || java -> applyJvm(it, extension)
+        else -> throw IllegalStateException("Impossible")
+      }
+    }
+  }
+
+  private fun applyJvm(
+    project: Project,
+    extension: WireExtension
+  ) {
+    val sourceConfiguration = project.configurations.create("wireSourceDependencies")
+
+    val sourcePaths = if (!extension.sourcePaths.isEmpty() || !extension.sourceTrees.isEmpty()) {
+      mergeDependencyPaths(project, extension.sourcePaths, extension.sourceTrees)
+    } else {
+      mergeDependencyPaths(project, setOf("src/main/proto"))
+    }
+    sourcePaths.forEach {
+      sourceConfiguration.dependencies.add(project.dependencies.create(it))
+    }
+
+    val protoConfiguration = project.configurations.create("wireProtoDependencies")
+
+    if (!extension.protoPaths.isEmpty() || !extension.protoTrees.isEmpty()) {
+      val allPaths = mergeDependencyPaths(project, extension.protoPaths, extension.protoTrees)
+      allPaths.forEach { path ->
+        protoConfiguration.dependencies.add(project.dependencies.create(path))
+      }
+    } else {
+      protoConfiguration.dependencies.addAll(sourceConfiguration.dependencies)
+    }
+
+    // all this point, all source and proto file references should be set up for Gradle to resolve.
+
+    val targets = mutableListOf<Target>()
+    val defaultBuildDirectory = "${project.buildDir}/generated/src/main/java"
+    val javaOutDirs = mutableListOf<String>()
+    val kotlinOutDirs = mutableListOf<String>()
+
+    val kotlinTarget = extension.kotlinTarget
+    val javaTarget = extension.javaTarget ?: if (kotlinTarget != null) null else JavaTarget()
+
+    javaTarget?.let { target ->
```

Comment:
two things:
1) one can have both a java and a kotlin target, to allow for incremental proto migration.
2) i separated the null check, since the first sets up the targets to be passed to the task, and the second links task dependencies.  open to suggestions on how to organize that better.

Reply:
didnt realize you could have both - makes sense!